### ISSUE TITLE
checkpoint与config文件不符

### ISSUE THREAD

**Issue by Artificialwhale** [HVision-NKU/Strip-R-CNN#33]
测试DOTA1.0数据集时，加载checkpoint “strip_rcnn_s_dota”会报：

unexpected key in source state_dict: roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial1.weight, roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial1.bias, roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial2.weight, roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial2.bias

missing keys in source state_dict: 
roi_head.bbox_head.reg_xy_wh_convs.1.strip_conv1.weight, roi_head.bbox_head.reg_xy_wh_convs.1.strip_conv1.bias, roi_head.bbox_head.reg_xy_wh_convs.1.strip_conv2.weight, roi_head.bbox_head.reg_xy_wh_convs.1.strip_conv2.bias, roi_head.bbox_head.reg_theta_fcs.0.weight, roi_head.bbox_head.reg_theta_fcs.0.bias, roi_head.bbox_head.reg_theta_fcs.1.weight, roi_head.bbox_head.reg_theta_fcs.1.bias

似乎缺少了 theta 的两个 fc 层，以及conv_spatial1/conv_spatial2不是StripHead 里的 StripBlock

**Comment by YXB-NKU** [comment]
是不是stripnet.py没更新，可以手动更改一下stripBlock里面参数的命名，和checkpoint一致就行

**Comment by changluei** [comment]
mmrotate/models/roi_heads/bbox_heads/reg_block.py 里的：

self.strip_conv1 = nn.Conv2d(...)
self.strip_conv2 = nn.Conv2d(...)
...
attn = self.strip_conv1(attn)
attn = self.strip_conv2(attn)

改成：

self.conv_spatial1 = nn.Conv2d(dim, dim, kernel_size=(1, 19), stride=1, padding=(0, 9), groups=dim)
self.conv_spatial2 = nn.Conv2d(dim, dim, kernel_size=(19, 1), stride=1, padding=(9, 0), groups=dim)
...
attn = self.conv_spatial1(attn)
attn = self.conv_spatial2(attn)

**Comment by Artificialwhale** [comment]
> mmrotate/models/roi_heads/bbox_heads/reg_block.py 里的：
> 
> self.strip_conv1 = nn.Conv2d(...) self.strip_conv2 = nn.Conv2d(...) ... attn = self.strip_conv1(attn) attn = self.strip_conv2(attn)
> 
> 改成：
> 
> self.conv_spatial1 = nn.Conv2d(dim, dim, kernel_size=(1, 19), stride=1, padding=(0, 9), groups=dim) self.conv_spatial2 = nn.Conv2d(dim, dim, kernel_size=(19, 1), stride=1, padding=(9, 0), groups=dim) ... attn = self.conv_spatial1(attn) attn = self.conv_spatial2(attn)

谢谢，unexpected key的问题解决了，但是仍然没有解决missing keys in source state_dict:
roi_head.bbox_head.reg_theta_fcs.0.weight, roi_head.bbox_head.reg_theta_fcs.0.bias, roi_head.bbox_head.reg_theta_fcs.1.weight, roi_head.bbox_head.reg_theta_fcs.1.bias

在StripHead中存在num_reg_theta_fcs=2
但是打印checkpoint中的head

roi_head.bbox_head.fc_cls.weight
roi_head.bbox_head.fc_cls.bias
roi_head.bbox_head.cls_fcs.0.weight
roi_head.bbox_head.cls_fcs.0.bias
roi_head.bbox_head.cls_fcs.1.weight
roi_head.bbox_head.cls_fcs.1.bias
roi_head.bbox_head.reg_xy_wh_convs.0.conv.weight
roi_head.bbox_head.reg_xy_wh_convs.0.conv.bias
roi_head.bbox_head.reg_xy_wh_convs.1.conv0.weight
roi_head.bbox_head.reg_xy_wh_convs.1.conv0.bias
roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial1.weight
roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial1.bias
roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial2.weight
roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial2.bias
roi_head.bbox_head.reg_xy_wh_convs.1.conv1.weight
roi_head.bbox_head.reg_xy_wh_convs.1.conv1.bias
roi_head.bbox_head.fc_reg_xy_wh.weight
roi_head.bbox_head.fc_reg_xy_wh.bias
roi_head.bbox_head.fc_reg_theta.weight
roi_head.bbox_head.fc_reg_theta.bias

发现模型确实缺少这俩fc层

**Comment by changluei** [comment]
> > mmrotate/models/roi_heads/bbox_heads/reg_block.py 里的：
> > self.strip_conv1 = nn.Conv2d(...) self.strip_conv2 = nn.Conv2d(...) ... attn = self.strip_conv1(attn) attn = self.strip_conv2(attn)
> > 改成：
> > self.conv_spatial1 = nn.Conv2d(dim, dim, kernel_size=(1, 19), stride=1, padding=(0, 9), groups=dim) self.conv_spatial2 = nn.Conv2d(dim, dim, kernel_size=(19, 1), stride=1, padding=(9, 0), groups=dim) ... attn = self.conv_spatial1(attn) attn = self.conv_spatial2(attn)
> 
> 谢谢，unexpected key的问题解决了，但是仍然没有解决missing keys in source state_dict: roi_head.bbox_head.reg_theta_fcs.0.weight, roi_head.bbox_head.reg_theta_fcs.0.bias, roi_head.bbox_head.reg_theta_fcs.1.weight, roi_head.bbox_head.reg_theta_fcs.1.bias
> 
> 在StripHead中存在num_reg_theta_fcs=2 但是打印checkpoint中的head
> 
> roi_head.bbox_head.fc_cls.weight roi_head.bbox_head.fc_cls.bias roi_head.bbox_head.cls_fcs.0.weight roi_head.bbox_head.cls_fcs.0.bias roi_head.bbox_head.cls_fcs.1.weight roi_head.bbox_head.cls_fcs.1.bias roi_head.bbox_head.reg_xy_wh_convs.0.conv.weight roi_head.bbox_head.reg_xy_wh_convs.0.conv.bias roi_head.bbox_head.reg_xy_wh_convs.1.conv0.weight roi_head.bbox_head.reg_xy_wh_convs.1.conv0.bias roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial1.weight roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial1.bias roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial2.weight roi_head.bbox_head.reg_xy_wh_convs.1.conv_spatial2.bias roi_head.bbox_head.reg_xy_wh_convs.1.conv1.weight roi_head.bbox_head.reg_xy_wh_convs.1.conv1.bias roi_head.bbox_head.fc_reg_xy_wh.weight roi_head.bbox_head.fc_reg_xy_wh.bias roi_head.bbox_head.fc_reg_theta.weight roi_head.bbox_head.fc_reg_theta.bias
> 
> 发现模型确实缺少这俩fc层

是的，我在复现的过程中也是这个问题，建议你采用backbone权重，然后在下游数据集上进行微调后获得完整权重。大概8卡需要12小时左右
